recipes/whisper-cpp/all/conanfile.py

```diff
@@ -114,13 +150,19 @@ def generate(self):
         if self.options.no_f16c:
             tc.variables["WHISPER_NO_F16C"] = True
 
+        if self.options.get_safe("with_openvino"):
+            tc.variables["WHISPER_OPENVINO"] = True
+            tc.variables["CMAKE_CXX_STANDARD"] = 11
```

Comment:
Recipes shouldn't hardcode a C++ standard. The approach to handle this is:
- The toolchain manages to define the C++ standard that the ``settings.compiler.cppstd`` define in the profile
- If this particular package has some constraint and it cannot be built with some cppstd (typically older ones, but sometimes also the newer ones), then it should raise a ``ConanInvalidConfiguration`` in the ``validate()`` method. (check also the ``validate_build()`` if there are different constraints to build than to use)

Reply:
PR is in draft.
Once https://github.com/conan-io/conan-center-index/pull/20782 is merged, that line will be removed.

Current problem is that whisper.cpp does not define any standard at all, while OpenVINO requires C++11, but does not expose it to its users. And as a result, whisper.cpp with openvino is compiled without even C++11, which leads to compilation issues.

Reply:
Ok, I see your point about C++ standard in toolchain. Looks like the WA is required only for old Conan 1.x?

Reply:
> And as a result, whisper.cpp with openvino is compiled without even C++11, which leads to compilation issues.


@ilya-lavrenov Could you please point these compilation issues with logs?

Reply:
https://c3i.jfrog.io/c3i/misc/logs/pr/20783/1-macos-clang/whisper-cpp/1.4.3//4137c3692310cc93d27861e3691668b109c494c1-build.txt

Reply:
Second example: https://github.com/conan-io/conan-center-index/pull/20785 where I removed explicit C++ standard from `test_package` and compilation has passed with `std_cxx_11`

Reply:
I've performed experiment with simple cmake project:
```cmake
cmake_minimum_required(VERSION 3.0.0 FATAL_ERROR)

set(CMAKE_CXX_STANDARD 17)

project(MyProj VERSION 0.1.0 LANGUAGES CXX)

add_executable(MyProj main.cpp)

target_compile_features(MyProj PRIVATE cxx_std_11)
```
And `make VERBOSE=1` prints:
```
[ 50%] Building CXX object CMakeFiles/MyProj.dir/main.cpp.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++   -std=gnu++17 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -MD -MT CMakeFiles/MyProj.dir/main.cpp.o -MF CMakeFiles/MyProj.dir/main.cpp.o.d -o CMakeFiles/MyProj.dir/main.cpp.o -c /Users/sandye51/Downloads/tests/main.cpp
[100%] Linking CXX executable MyProj
```
If I comment explicit setting `CMAKE_CXX_STANDARD` to 17, then:
```
[ 50%] Building CXX object CMakeFiles/MyProj.dir/main.cpp.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++   -std=gnu++11 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -MD -MT CMakeFiles/MyProj.dir/main.cpp.o -MF CMakeFiles/MyProj.dir/main.cpp.o.d -o CMakeFiles/MyProj.dir/main.cpp.o -c /Users/sandye51/Downloads/tests/main.cpp
[100%] Linking CXX executable MyProj
```
So, conclusion: `target_compile_features` defines requirements for targets, but does not override `CXX_STANDARD`, which is set either globally or via `set_target_properties` for specific target.

Reply:
Hi @memsharded @uilianries 
Please, provide your opinion on my input above

Reply:
In Conan 2.0, ``check_min_cppstd()`` which is what should be used in ``validate()`` method to require C++11 as input, will raise if ``compiler.cppstd`` is not defined.
So ``compiler.cppstd`` is not mandated by Conan, but in practice, recipes mandate it being defined and with the correct minimum value. The ``conan profile detect`` command also assigns a ``compiler.cppstd`` value by default and ConanCenter profiles define it too. It is expected to always be defined in Conan 2.0 profiles.

So this is the expected behavior in Conan 2.0, which would make unnecessary adding these changes to recipe. Having a declarative ``compiler.cppstd`` has proven more robust and being able to fail faster when the user doesn't provide a minimum one works better than changing the user c++ standard  behind the scenes because some transitive dependency has a different requirement that is propagated. If you are going to build your code with C++17, you should be declaring it, not happening because some dependency injected the requirement.

We could argue about Conan 1.X, but the plan is to freeze updates to ConanCenter for 1.X in the following months (not breaking any 1.X users, but stop updating and building packages, they will keep working with the existing ones, not being removed). So we'd prefer to not add these to recipes in ConanCenter, but follow the ``validate()`` recommended approach.

Reply:
My comment is also valid for https://github.com/conan-io/conan-center-index/pull/20782. The overall design is:
- ``compiler.cppstd`` is defined declarative, and "imperatively" in profiles
- Packages should raise if the ``compiler.cppstd`` is not enough for their needs. This also makes the definition of ``compiler.cppstd`` mandatory in Conan 2.0 for these packages to work.
- Package shouldn't propagate cppstd requirements to their consumers under the hood, via cmake_modules, properties, compiler flags, or any other means

Reply:
Thank you for your answer, @memsharded 

> Package shouldn't propagate cppstd requirements to their consumers under the hood

And https://github.com/conan-io/conan-center-index/pull/20782 does not do it. As proven by example above, `target_compile_features` does not change externally set `compiler.cppstd` (read as "does not break Conan 2.0 users"). 

> We could argue about Conan 1.X, but the plan is to freeze updates to ConanCenter for 1.X

But this is just a plan, which may not happen. "Months" is too long and we need solution right now, because in current situation I cannot merge this PR, because:
- I cannot explicitly set `CMAKE_CXX_STANDARD`, because it's prohibited by Conan.
- I also cannot fix OpenVINO recipe, because as I understand you don't recommend to add `target_compile_features`, because of "freeze updates to ConanCenter for 1.X".

> My comment is also valid for

I don't think the comment is valid, because I'm not talking about Conan 2.0, which is not affected as proven by example below. 

I want to hear what you recommend in this particular case, but not general considerations about Conan policies regarding C++ standards.